Autopilot/Path Manager/MPL3115A2.c

```diff
@@ -30,36 +33,80 @@ char initAltimeter() {
 #endif
     }
     return altimeterConnected;
+    // set pressure 
 }
 
-void calibrateAltimeter(long int pressure){
-    int msb = pressure >> 9;
-    int lsb = (pressure >> 1)& 255;
-    int offset = 0;
-    sendMessage(I2C_SLAVE_ADDRESS, BAR_IN_MSB, (char*)&msb, 1, WRITE);
-    sendMessage(I2C_SLAVE_ADDRESS, BAR_IN_LSB, (char*)&lsb, 1, WRITE);
-    sendMessage(I2C_SLAVE_ADDRESS, OFF_H, (char*)&offset, 1, WRITE);
+void calibrateAltimeter(int altitudeOffset){
+    if (altitudeOffset < -128 || altitudeOffset > 128){
+        error("Invalid altitude offset!");
+        // the range of altitude offset should be within -128 to 128
+    }
+    
+    else{
+        switchMode(1); //Barometer mode
+        float pressureFloat = 0.0;
+        //There's slight delay on pressure display, wait until pressure is not zero
+        while(pressureFloat <= 0.1){
+            pressureFloat = getPressure();
+        }
+        uint32_t pressure = (int) pressureFloat;
+        switchMode(0); //Altimeter mode
+
+        // set ground pressure
+        pressure = pressure << 1; //Value is input in 2 Pascal units
+        int msb = pressure >> 8;
+        int lsb = pressure & 0xFF;
+
+        sendMessage(I2C_SLAVE_ADDRESS, BAR_IN_MSB, (char*)&msb, 1, WRITE);
+        sendMessage(I2C_SLAVE_ADDRESS, BAR_IN_LSB, (char*)&lsb, 1, WRITE);
+        sendMessage(I2C_SLAVE_ADDRESS, OFF_H, (char*)&altitudeOffset, 1, WRITE);
+    }
 }
 
 float getAltitude() {
     if (altimeterConnected && (sendMessage(I2C_SLAVE_ADDRESS, DATA_READY_REGISTER, 0, 0, READ) & 0x08)) {
-        unsigned char pre_msb = sendMessage(I2C_SLAVE_ADDRESS, ALTITUDE_MSB_REGISTER, 0, 0, READ);
-        unsigned char pre_csb = sendMessage(I2C_SLAVE_ADDRESS, ALTITUDE_CSB_REGISTER, 0, 0, READ);
-        unsigned char pre_lsb = sendMessage(I2C_SLAVE_ADDRESS, ALTITUDE_LSB_REGISTER, 0, 0, READ);
-
-       /* 
-        //In barometer mode, this is the way to read current pressure by using the altimeter 
-        unsigned long int msb = pre_msb;// convert into long int 32bit
-        msb = msb << 10;// shit left 10
-        unsigned int csb = pre_csb << 2;
-        float lsb = (pre_lsb >> 4) / 4.0;
-        lastKnownAltitude = (float) (msb | csb) + lsb;
-        */
+        uint8_t pre_msb = sendMessage(I2C_SLAVE_ADDRESS, ALTITUDE_MSB_REGISTER, 0, 0, READ);
+        uint8_t pre_csb = sendMessage(I2C_SLAVE_ADDRESS, ALTITUDE_CSB_REGISTER, 0, 0, READ);
+        uint8_t pre_lsb = sendMessage(I2C_SLAVE_ADDRESS, ALTITUDE_LSB_REGISTER, 0, 0, READ);
 
         int msb = pre_msb << 8;
-        unsigned char csb = pre_csb;
+        uint8_t csb = pre_csb;
         float lsb = (pre_lsb >> 4)/16.0;
-        lastKnownAltitude = (float) (msb | csb) + lsb - altimeterOffset;
+        lastKnownAltitude = (float) (msb | csb) + lsb;
+ 
     }
     return lastKnownAltitude;
+}
+
+// toBarometer = 1 -> switch to Barometer
+// toBarometer = 0 -> switch to Altimeter
+void switchMode(int toBarometer) {
```

Comment:
a better way of doing this would be to declare an enum at the top and use that instead of this. Would make the code clearer. For example:
```c
typedef enum {
   ALTMETER,
   BAROMETER
} ALTMETER_MODE;
```
I think that syntax is right